pkg/downwardmetrics/virtio-serial/server.go

```diff
@@ -0,0 +1,276 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2023 Red Hat, Inc.
+ *
+ */
+
+package virtio_serial
+
+import (
+	"bufio"
+	"encoding/xml"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/textproto"
+	"net/url"
+	"strings"
+	"syscall"
+	"time"
+
+	"kubevirt.io/client-go/log"
+
+	"kubevirt.io/kubevirt/pkg/downwardmetrics"
+	"kubevirt.io/kubevirt/pkg/downwardmetrics/vhostmd/api"
+	"kubevirt.io/kubevirt/pkg/util"
+	"kubevirt.io/kubevirt/pkg/virt-handler/isolation"
+
+	diskutils "kubevirt.io/kubevirt/pkg/ephemeral-disk-utils"
+	metricsScraper "kubevirt.io/kubevirt/pkg/monitoring/domainstats/downwardmetrics"
+)
+
+const connectAttempts = 6
+const standardLauncherSocketPath = "/run/kubevirt/sockets/launcher-sock"
+
+var (
+	invalidRequest = []byte("INVALID REQUEST\n\n")
+	emptyMetrics   = []byte("<metrics><!-- host metrics not available --><!-- VM metrics not available --></metrics>")
+)
+
+func RunDownwardMetricsVirtioServer(nodeName string, stop chan struct{}) error {
+	report, err := newMetricsReporter(nodeName)
+	if err != nil {
+		return err
+	}
+
+	server := downwardMetricsServer{
+		connectAttempts:    connectAttempts,
+		virtioSerialSocket: downwardmetrics.DownwardMetricsChannelSocket,
+		report:             report,
+	}
+	return server.start(stop)
+}
+
+type metricsReporter func() (*api.Metrics, error)
+
+func newMetricsReporter(nodeName string) (metricsReporter, error) {
+	exists, err := diskutils.FileExists(standardLauncherSocketPath)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, errors.New("virt-launcher socket not found")
+	}
+
+	isolationDetector := isolation.NewSocketBasedIsolationDetector(util.VirtShareDir)
+	scraper := metricsScraper.NewScraper(nodeName, isolationDetector)
+
+	return func() (*api.Metrics, error) {
+		return scraper.Report(standardLauncherSocketPath)
+	}, nil
+}
+
+type downwardMetricsServer struct {
+	connectAttempts    uint
+	virtioSerialSocket string
+	report             metricsReporter
+}
+
+func (s *downwardMetricsServer) start(stop chan struct{}) error {
+	conn, err := connect(s.virtioSerialSocket, s.connectAttempts)
+	if err != nil {
+		return err
+	}
+
+	go s.serve(conn, stop)
+	return nil
+}
+
+func (s *downwardMetricsServer) serve(conn net.Conn, stop chan struct{}) {
+	defer func(conn net.Conn) {
+		err := conn.Close()
+		if err != nil {
+			log.Log.Reason(err).Warning("virtio-serial failed to close the connection")
+		}
+	}(conn)
+
+	type reqResult struct {
+		request string
+		err     error
+	}
+	newRequest := make(chan reqResult)
```

Comment:
Could it make sense to make this a buffered channel?

Reply:
it's not needed, for two reasons, first the UDS already has an internal buffer and last since it's a virtio-serial device we can server just one request at a time